### DISCUSSION TITLE
Allow verbs in the specification

### DISCUSSION THREAD

**Discussion by LSWdigital** [cooklang/spec#56] [Ideas]
I have built a similar (toy) language for recipes in the past, and noted that such a language could be interpreted by a robot.

For this to take place, verbs should be available to the language, similar to the @, #, and ~ nouns. $ may be a sensible character to use.

The "standard library" should not provide any verbs (just as no @ or # nouns are provided).

**Comment by dubadub** [comment]
Hi @LSWdigital I'm not exactly sure what you mean. Could you please give an example how verb is used?

**Comment by LSWdigital** [comment]
An example is given in the readme

> Mash \@potato{2%kg} until smooth

In principle, a robot chef could mash potatoes. It could provide an implementation cooklang with a builtin $mash{} in order to do this.

> $mash \@potato{2%kg} until smooth.

This would admittedly not be particularly useful for the main intended use of the language, but I think as an optional feature, it could keep the potential open as the language is developed.

**Comment by dubadub** [comment]
@LSWdigital thanks for sharing the example. It's still not clear for me what is the use case and how this will be used in the end. Also, I assume we need to express a link between the verb and ingredient to make any use of it, right?

**Reply by Silthanis** [reply]
Not the original requester, but my use case is to output [Cooking for Engineers](https;//cookingforengineers.com) style recipe tables:

<table><tr><td colspan=7 align=center>Butter and flour a loaf pan</td></tr><tr><td colspan=7 align=center>Preheat oven to 350&#176;F (170&#176;C)</td></tr><tr><td><span itemprop="ingredient" itemscope itemtype="http://data-vocabulary.org/RecipeIngredient"><span itemprop="amount">2 large (250 g)</span> <span itemprop="name">ripe bananas</span></span></td><td>mash</td><td rowspan=4 class="vertical">mash until smooth</td><td rowspan=10 class="vertical">fold</td><td rowspan=10 class="vertical">bake 350&#176;F (170&#176;C) 55 min.</td><td rowspan=10 class="vertical">cool 10 min. in pan</td><td rowspan=10 class="vertical">cool on wire rack</td></tr><tr><td><span itemprop="ingredient" itemscope itemtype="http://data-vocabulary.org/RecipeIngredient"><span itemprop="amount">6 Tbs. (90 mL)</span> <span itemprop="name">butter</span></span></td><td>melt</td></tr><tr><td><span itemprop="ingredient" itemscope itemtype="http://data-vocabulary.org/RecipeIngredient"><span itemprop="amount">1 tsp. (5 mL)</span> <span itemprop="name">vanilla extract</span></span></td><td class="righthide"></td></tr><tr><td><span itemprop="ingredient" itemscope itemtype="http://data-vocabulary.org/RecipeIngredient"><span itemprop="amount">2 large</span> <span itemprop="name">eggs</span></span></td><td>lightly beat</td></tr><tr><td><span itemprop="ingredient" itemscope itemtype="http://data-vocabulary.org/RecipeIngredient"><span itemprop="amount">1-1/3 cups (167 g)</span> <span itemprop="name">all-purpose flour</span></span></td><td rowspan=5 class="vertical">whisk</td><td rowspan=6 class="righthide"></td></tr><tr><td><span itemprop="ingredient" itemscope itemtype="http://data-vocabulary.org/RecipeIngredient"><span itemprop="amount">2/3 cup (130 g)</span> <span itemprop="name">sugar</span></span></td></tr><tr><td><span itemprop="ingredient" itemscope itemtype="http://data-vocabulary.org/RecipeIngredient"><span itemprop="amount">1/2 tsp. (2.3 g)</span> <span itemprop="name">baking soda</span></span></td></tr><tr><td><span itemprop="ingredient" itemscope itemtype="http://data-vocabulary.org/RecipeIngredient"><span itemprop="amount">1/4 tsp. (1.2 g)</span> <span itemprop="name">baking powder</span></span></td></tr><tr><td><span itemprop="ingredient" itemscope itemtype="http://data-vocabulary.org/RecipeIngredient"><span itemprop="amount">1/2 tsp. (3 g)</span> <span itemprop="name">salt</span></span></td></tr><tr><td><span itemprop="ingredient" itemscope itemtype="http://data-vocabulary.org/RecipeIngredient"><span itemprop="amount">1/2 cup (70 g)</span> <span itemprop="name">chopped walnuts</span></span></td><td class="righthide"></td></tr></table>

**Comment by potato** [comment]
The fact that i'm mentioned in a repo i never heard about is hilarious, i kinda like the markup ;)
Great project btw!

**Comment by dubadub** [comment]
@potato haha, great username!

**Comment by keithamus** [comment]
I have built a style guide for my recipes. My recipes all use very dense/terse style of instruction where every step has to start with a verb, which is from a fixed list in the style guide. Admittedly this is a little pednantic/opinionated (though perhaps this lines up with the spirit of cooklang anyway). It's useful for a few reasons:

1. Enforcing a verb at the start of every step allows you to quickly scan a recipe and get a lot of data, e.g a 4 step recipe with "Chop, Oil, Fry, Assemble" tells you its likely simple, quick and low effort, while "Mix, Knead, Transfer, Cover, Prove, Divide, Line, Boil, Bake, Cool" tells you this is am involved batch recipe with likely multiple rest periods.
2. Having a step of "Take some oil and rub down a pan, covering it with the oil" is a bit verbose when you can just say "Coat", but if you're going to be this term a glossary to define and disambiguate these terms is helpful. e.g. "Oil a pan" is terse but could mean either "coat the whole pan with a thing layer fo oil" or could mean "drizzle a bit of oil in the pan". My style guide has these as two different verbs "Coat" and "Oil".
3. Some languages/locales have different terms for different things. I'm from the UK, we don't "broil" things, we "grill" them. My style guide lists synonyms so I can easily convert terms if I'm importing/rewriting a recipe.
4. Recipes can have a ton of synonyms for techniques, e.g. "hydrate", "rehydrate", "soak". Having a canonical term for this can be useful.
5. Sometimes recipes can be imprecise or verbose ("Cut the potato into cubes"). My style guide bans the word Cut, instead requiring a cutting verb like "Halve, Quarter, Cube, Dice, Julienne, Chiffonade". This front-loads the step as one word rather than a whole sentence (`Cube @potatoes{}`, `Chiffonade @parsley{}`).
6. (My style guide doesn't do this but) I could easily imagine some advanced techniques (e.g. Velvet, Chiffonade) could have explanations or maybe a video link so new readers (or if you're sharing a recipe) can get up to speed with those terms.

So to give you a peek at style guide, it has sections like this:

```
### Surface Preparation

Preparing pans, trays, and moulds.

| Verb   | Meaning                                          |
| ------ | ------------------------------------------------ |
| Grease | Apply fat to a surface to prevent sticking       |
| Brush  | Apply a thin coat of liquid using a pastry brush |
| Coat   | Cover a surface or ingredient with a layer       |
| Oil    | Apply oil to a surface                           |
| Flour  | Dust a surface with flour to prevent sticking    |
| Butter | Apply butter to a surface                        |
| Spray  | Apply oil from a spray bottle                    |
| Rub    | Apply by pressing and moving across a surface    |

```

My style guide has around 100 verbs like this. Some tables also include synonyms so I can catch words I might come across in a recipe, and normalise them to my set.

It would be very useful for me if I could encode this into the tool itself, rather than having a separate style guide that I simply follow and reference.

**Reply by gvlx** [reply]
Integrating this table into the system, could help on the editor and the server
